virtwho/config.py

```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
+        elif len(self._values[key]) == 0:
+            result = ('warning', '%s cannot be empty, using default' % key)
+        return result
+
+    def validate(self, validation_messages=None):
+        validation_messages = super(GlobalSection, self).validate(validation_messages)
+        to_reset = set()
+        # Validate those keys that need to be validated
+        for key in set(self._unvalidated_keys):
+            error = None
+            # If boolean
+            if key in ['debug', 'oneshot', 'print', 'log_per_config', 'background']:
+                error = self._validate_str_to_bool(key)
+            elif key == 'configs':
+                error = self._validate_configs()
+            elif key in ['reporter_id', 'log_file', 'log_dir']:
+                error = self._validate_non_empty_string(key)
+            elif key == 'interval':
+                error = self._validate_interval()
+            else:
+                validation_messages.append(('warning', 'Ignoring unknown configuration option '
+                                                       '"%s"' % key))
+                del self._values[key]
+            if error is not None:
+                validation_messages.append(error)
+                if key not in ['interval']:  # Special cases not reset to default on failure
+                    to_reset.add(key)
+                self._invalid_keys.add(key)
+            self._unvalidated_keys.remove(key)
+
+        for key in to_reset:
+            if self.has_default(key):
+                self._values[key] = self.defaults[key]
+                if key in self._invalid_keys:
+                    self._invalid_keys.remove(key)
+
+        if self._values['print']:
+            self._values['oneshot'] = True
+
+        self._update_state()
+
+        return validation_messages
+
+# String representations of all the default configuration for virt-who
+DEFAULTS = {
+    VW_GLOBAL: GlobalSection.get_defaults(),
+    VIRTWHO_ENV_CLI_SECTION_NAME: {
+        'smtype': 'sam',
+    },
+}
+
+
+class EffectiveConfig(collections.MutableMapping):
+    """
+    This object represents the total configuration of virt-who including all global parameters
+    and all sections that define a source or destination.
+    """
+
+    class State(object):
+        VALID = 0
+        INVALID = 1
+        UNKNOWN = -1
+        NEEDS_VALIDATION = -2
+
+    __marker = object()
+
+    def __iter__(self):
+        return iter(self._sections)
+
+    def __delitem__(self, key):
+        if key in self:
+            del self[key]
+        else:
+            raise KeyError("Unable to delete nonexistant section '%s'" % key)
+
+    def __setitem__(self, key, value):
+        self._sections[key] = value
+
+    def __len__(self):
+        return len(self._sections)
+
+    def __getitem__(self, key):
+        return self._sections[key]
+
+    def __contains__(self, item):
+        return item in self._sections
+
+    def __init__(self, env_args, cli_args):
```

Comment:
Would the preference to be have a module level method which creates an instance of the EffectiveConfig object?